src/classad/exprTree.cpp

```diff
@@ -353,6 +331,49 @@ SetRootScope( )
     return;
 }
 
+// add an ExprTree to the cache of things to delete after evaluation is complete
+void EvalState::AddToDeletionCache(ExprTree * tree)
+{
+	// protect against double-free by only adding things that are not already in the collection
+	auto found = std::find(cache_to_delete.begin(), cache_to_delete.end(), tree);
+	if (found == cache_to_delete.end()) {
+		cache_to_delete.push_back(tree);
+	}
+}
+
+// look for the given ExprTree pointer in the cache, and if it is found
+// remove it from the cache and return true.
+bool EvalState::TakeFromDeletionCache(ExprTree * tree) {
+	auto last = std::remove(cache_to_delete.begin(), cache_to_delete.end(), tree);
+	if (last != cache_to_delete.end()) {
+		cache_to_delete.erase(last, cache_to_delete.end());
+		return true;
+	}
+	return false;
+}
+
+// if a Value currently has a pointer that is also in the evaluation cache
+// transfer ownership of the pointer to the Value (removing it from the cache)
+// and convert the Value to a counted-pointer type, thus making the value self-contained.
+// This is an optimization for the case where evaluation creates a temporary ExprTree
+// that is the result of evaluation.  We can avoid deep-copying the tree and then
+// turning around and deleting the original.
+bool EvalState::GivePtrToValue(Value & val)
```

Comment:
This feels like it should just be in the body of Value::MakeSelfContained(). Then, Value doesn't have to make EvalState a friend.